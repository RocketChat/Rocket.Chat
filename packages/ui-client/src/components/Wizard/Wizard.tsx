import { memo, type ReactNode } from 'react';

import { WizardContext } from './WizardContext';
import type { WizardAPI } from './WizardContext';

type WizardProps = {
	api: WizardAPI;
	children: ReactNode;
};

const Wizard = ({ children, api }: WizardProps) => (
	<WizardContext.Provider value={api}>
		<div className='steps-wizard'>{children}</div>
	</WizardContext.Provider>
);

export default memo(Wizard);
