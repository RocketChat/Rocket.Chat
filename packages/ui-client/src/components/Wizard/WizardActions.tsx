import { Box, ButtonGroup } from '@rocket.chat/fuselage';
import type { ReactNode } from 'react';

type WizardActionsProps = {
	annotation?: string;
	children: ReactNode;
};

const WizardActions = ({ annotation, children }: WizardActionsProps) => (
	<Box className='steps-wizard-footer' pbs={24} display='flex' alignItems='center' justifyContent='end'>
		{annotation ? (
			<Box mie='auto' fontScale='c1' color='annotation'>
				{annotation}
			</Box>
		) : null}

		<ButtonGroup>{children}</ButtonGroup>
	</Box>
);

export default WizardActions;
