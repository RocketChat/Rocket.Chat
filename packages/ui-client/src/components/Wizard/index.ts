export { default } from './Wizard';
export { default as Wizard } from './Wizard';
export { default as WizardContent } from './WizardContent';
export { default as WizardTabs } from './WizardTabs';
export { default as WizardActions } from './WizardActions';
export { default as WizardBackButton } from './WizardBackButton';
export { default as WizardNextButton } from './WizardNextButton';
export { default as StepsLinkedList } from './lib/StepsLinkedList';
export { default as StepNode } from './lib/StepNode';
export * from './lib/StepNode';
export * from './WizardContext';
export * from './useWizardContext';
export * from './useWizard';
export * from './mocks/createMockWizardApi';
