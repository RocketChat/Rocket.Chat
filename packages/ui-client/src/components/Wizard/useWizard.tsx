import { useEffectEvent } from '@rocket.chat/fuselage-hooks';
import { useMemo, useState } from 'react';

import type { WizardAPI } from './WizardContext';
import type { StepMetadata } from './lib/StepNode';
import type StepNode from './lib/StepNode';
import StepsLinkedList from './lib/StepsLinkedList';

type UseWizardProps = {
	steps: StepMetadata[];
};

/**
 * Custom hook to manage the state and navigation of a wizard.
 * It provides methods to register steps, navigate between them, and manage their state.
 *
 * @param {UseWizardProps} props - The properties for the wizard.
 * @returns {WizardAPI} The API for managing the wizard state and navigation.
 */
export const useWizard = ({ steps: stepsMetadata }: UseWizardProps) => {
	const [steps] = useState(new StepsLinkedList(stepsMetadata));
	const [currentStep, setCurrentStep] = useState<StepNode | null>(steps.head);

	/**
	 * Registers a new step in the wizard.
	 * If a step with the same ID already exists, it updates the existing step.
	 */
	const register = useEffectEvent((stepMetadata: StepMetadata) => {
		const step = steps.append(stepMetadata);
		return () => steps.remove(step.id);
	});

	/**
	 * Navigates to a specific step in the wizard.
	 * If the step is disabled, it does nothing.
	 *
	 * @param {StepNode} step - The step to navigate to.
	 */
	const goTo = useEffectEvent(async (step: StepNode) => {
		if (step.disabled) {
			return;
		}

		setCurrentStep(step);
	});

	/**
	 * Navigates to the next step in the wizard.
	 * If there is no next step, it does nothing.
	 */
	const next = useEffectEvent(async () => {
		if (!currentStep?.next) {
			return;
		}

		steps.enableStep(currentStep.next);
		goTo(currentStep.next);
	});

	/**
	 * Navigates to the previous step in the wizard.
	 * If there is no previous step, it does nothing.
	 */
	const previous = useEffectEvent(async () => {
		if (!currentStep?.prev) {
			return;
		}

		steps.enableStep(currentStep.prev);
		goTo(currentStep.prev);
	});

	/**
	 * Resets the next steps in the wizard.
	 * It disables all steps that come after the current step.
	 */
	const resetNextSteps = useEffectEvent(() => {
		if (!currentStep) {
			return;
		}

		let step = currentStep;
		while (step.next) {
			steps.disableStep(step.next);
			step = step.next;
		}
	});

	return useMemo<WizardAPI>(
		() => ({
			steps,
			register,
			currentStep,
			next,
			previous,
			goTo,
			resetNextSteps,
		}),
		[currentStep, goTo, next, previous, register, steps, resetNextSteps],
	);
};
