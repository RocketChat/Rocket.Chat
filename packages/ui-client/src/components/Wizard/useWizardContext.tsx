import { useContext } from 'react';

import { WizardContext } from './WizardContext';

export const useWizardContext = () => {
	const context = useContext(WizardContext);

	if (!context) {
		throw new Error('useWizardContext must be used within a WizardProvider');
	}

	return context;
};
