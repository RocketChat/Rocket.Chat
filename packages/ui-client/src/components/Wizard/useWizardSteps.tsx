import { useCallback, useRef, useSyncExternalStore } from 'react';

import type StepNode from './lib/StepNode';
import type StepsLinkedList from './lib/StepsLinkedList';

/**
 * Custom hook to manage the state of wizard steps.
 * It uses a linked list to store the steps and provides a way to subscribe to changes.
 *
 * @param {StepsLinkedList} list - The linked list containing the steps.
 * @returns {StepNode[]} The current state of the steps.
 */
export const useWizardSteps = (list: StepsLinkedList) => {
	const stateRef = useRef<StepNode[]>([]);

	const getSnapshot = useCallback(() => stateRef.current, []);

	const subscribe = useCallback(
		(onStoreChange: () => void) => {
			stateRef.current = list.toArray();

			return list.on('stateChanged', (): void => {
				stateRef.current = list.toArray();
				onStoreChange();
			});
		},
		[list],
	);

	return useSyncExternalStore(subscribe, getSnapshot);
};
