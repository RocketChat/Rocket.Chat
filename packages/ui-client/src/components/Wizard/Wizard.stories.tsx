import { Box, Button, States, StatesIcon, StatesTitle } from '@rocket.chat/fuselage';
import type { Meta, StoryFn } from '@storybook/react';

import Wizard from './Wizard';
import WizardActions from './WizardActions';
import WizardBackButton from './WizardBackButton';
import WizardContent from './WizardContent';
import WizardNextButton from './WizardNextButton';
import WizardTabs from './WizardTabs';
import { useWizard } from './useWizard';

export default {
	title: 'Components/Wizard',
	component: Wizard,
	subcomponents: {
		WizardActions,
		WizardBackButton,
		WizardContent,
		WizardNextButton,
		WizardTabs,
	},
	parameters: {
		layout: 'centered',
	},
	decorators: [
		(Story) => (
			<Box>
				<Story />
			</Box>
		),
	],
} satisfies Meta<typeof Wizard>;

const WizardExample = ({ ordered = false }: { ordered?: boolean }) => {
	const wizardApi = useWizard({
		steps: [
			{ id: 'first-step', title: 'First step' },
			{ id: 'second-step', title: 'Second step' },
			{ id: 'third-step', title: 'Third step' },
		],
	});

	return (
		<Wizard api={wizardApi}>
			<WizardTabs ordered={ordered} />

			<WizardContent id='first-step'>
				<Box width='100%' height='100%' pbs={24}>
					<States>
						<StatesIcon name='magnifier' />
						<StatesTitle>First step</StatesTitle>
					</States>
				</Box>

				<WizardActions>
					<WizardNextButton />
				</WizardActions>
			</WizardContent>
			<WizardContent id='second-step'>
				<Box width='100%' height='100%' pbs={24}>
					<States>
						<StatesIcon name='magnifier' />
						<StatesTitle>Second step</StatesTitle>
					</States>
				</Box>

				<WizardActions>
					<WizardBackButton />
					<WizardNextButton />
				</WizardActions>
			</WizardContent>
			<WizardContent id='third-step'>
				<Box width='100%' height='100%' pbs={24}>
					<States>
						<StatesIcon name='magnifier' />
						<StatesTitle>Third step</StatesTitle>
					</States>
				</Box>

				<WizardActions>
					<Button primary onClick={() => window.alert('Finished!')}>
						Finish
					</Button>
				</WizardActions>
			</WizardContent>
		</Wizard>
	);
};

export const BasicWizard: StoryFn<typeof Wizard> = () => <WizardExample />;

export const OrderedTabsWizard: StoryFn<typeof Wizard> = () => <WizardExample ordered />;
