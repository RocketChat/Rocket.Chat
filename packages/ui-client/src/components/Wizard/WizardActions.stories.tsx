import { Box } from '@rocket.chat/fuselage';
import type { Meta, StoryFn } from '@storybook/react';
import { ComponentProps } from 'react';

import Wizard from './Wizard';
import WizardActions from './WizardActions';
import WizardBackButton from './WizardBackButton';
import WizardNextButton from './WizardNextButton';
import { useWizard } from './useWizard';

export default {
	title: 'Components/Wizard/WizardActions',
	component: WizardActions,
	decorators: (Story) => {
		const wizardApi = useWizard({
			steps: [
				{ id: 'first-step', title: 'First step' },
				{ id: 'second-step', title: 'Second step' },
				{ id: 'third-step', title: 'Third step' },
			],
		});

		return (
			<Box width={600}>
				<Wizard api={wizardApi}>
					<Story />
				</Wizard>
			</Box>
		);
	},
	parameters: {
		layout: 'centered',
	},
} satisfies Meta<typeof WizardActions>;

const Template: StoryFn<ComponentProps<typeof WizardActions>> = (args) => (
	<WizardActions {...args}>
		<WizardBackButton />
		<WizardNextButton />
	</WizardActions>
);

export const Default = Template.bind({});
Default.args = {};

export const WithAnnotation = Template.bind({});
WithAnnotation.args = {
	annotation: 'This is a sample annotation',
};
