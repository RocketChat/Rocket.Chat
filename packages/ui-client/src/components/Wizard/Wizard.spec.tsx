import { composeStories } from '@storybook/react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import * as stories from './Wizard.stories';

const testCases = Object.values(composeStories(stories)).map((Story) => [Story.storyName || 'Story', Story]);

test.each(testCases)(`renders %s without crashing`, async (_storyname, Story) => {
	const { baseElement } = render(<Story />);
	expect(baseElement).toMatchSnapshot();
});

test.each(testCases)('%s should have no a11y violations', async (_storyname, Story) => {
	const { container } = render(<Story />);

	const results = await axe(container);
	expect(results).toHaveNoViolations();
});
