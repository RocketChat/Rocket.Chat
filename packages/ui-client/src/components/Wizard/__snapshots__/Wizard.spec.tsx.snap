// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`renders BasicWizard without crashing 1`] = `
<body>
  <div>
    <div
      class="rcx-box rcx-box--full"
    >
      <div
        class="steps-wizard"
      >
        <div
          class="rcx-box rcx-box--full rcx-tabs--with-divider rcx-tabs rcx-css-1tbw8nv"
        >
          <div
            class="rcx-box rcx-box--full rcx-tabs__scroll-box"
          >
            <div
              class="rcx-box rcx-box--full rcx-tabs__wrapper"
              role="tablist"
            >
              <button
                aria-selected="true"
                class="rcx-box rcx-box--full rcx-tabs__item--selected rcx-tabs__item rcx-css-5dmxf6"
                role="tab"
              >
                First step
              </button>
              <button
                aria-selected="false"
                class="rcx-box rcx-box--full rcx-tabs__item--disabled rcx-tabs__item rcx-css-5dmxf6"
                role="tab"
              >
                Second step
              </button>
              <button
                aria-selected="false"
                class="rcx-box rcx-box--full rcx-tabs__item--disabled rcx-tabs__item rcx-css-5dmxf6"
                role="tab"
              >
                Third step
              </button>
            </div>
          </div>
        </div>
        <div>
          <div
            class="rcx-box rcx-box--full rcx-css-4q8ri8"
          >
            <div
              class="rcx-states"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-states__icon rcx-icon--name-magnifier rcx-icon rcx-css-d884y5"
              >
                
              </i>
              <h3
                class="rcx-states__title"
              >
                First step
              </h3>
            </div>
          </div>
          <div
            class="rcx-box rcx-box--full steps-wizard-footer rcx-css-7vikfg"
          >
            <div
              class="rcx-button-group rcx-button-group--align-start"
              role="group"
            >
              <button
                class="rcx-box rcx-box--full rcx-button--primary rcx-button rcx-button-group__item"
                type="button"
              >
                <span
                  class="rcx-button--content"
                >
                  Next
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders OrderedTabsWizard without crashing 1`] = `
<body>
  <div>
    <div
      class="rcx-box rcx-box--full"
    >
      <div
        class="steps-wizard"
      >
        <div
          class="rcx-box rcx-box--full rcx-tabs--with-divider rcx-tabs rcx-css-1tbw8nv"
        >
          <div
            class="rcx-box rcx-box--full rcx-tabs__scroll-box"
          >
            <div
              class="rcx-box rcx-box--full rcx-tabs__wrapper"
              role="tablist"
            >
              <button
                aria-selected="true"
                class="rcx-box rcx-box--full rcx-tabs__item--selected rcx-tabs__item rcx-css-5dmxf6"
                role="tab"
              >
                1. First step
              </button>
              <button
                aria-selected="false"
                class="rcx-box rcx-box--full rcx-tabs__item--disabled rcx-tabs__item rcx-css-5dmxf6"
                role="tab"
              >
                2. Second step
              </button>
              <button
                aria-selected="false"
                class="rcx-box rcx-box--full rcx-tabs__item--disabled rcx-tabs__item rcx-css-5dmxf6"
                role="tab"
              >
                3. Third step
              </button>
            </div>
          </div>
        </div>
        <div>
          <div
            class="rcx-box rcx-box--full rcx-css-4q8ri8"
          >
            <div
              class="rcx-states"
            >
              <i
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-states__icon rcx-icon--name-magnifier rcx-icon rcx-css-d884y5"
              >
                
              </i>
              <h3
                class="rcx-states__title"
              >
                First step
              </h3>
            </div>
          </div>
          <div
            class="rcx-box rcx-box--full steps-wizard-footer rcx-css-7vikfg"
          >
            <div
              class="rcx-button-group rcx-button-group--align-start"
              role="group"
            >
              <button
                class="rcx-box rcx-box--full rcx-button--primary rcx-button rcx-button-group__item"
                type="button"
              >
                <span
                  class="rcx-button--content"
                >
                  Next
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
